core/src/multi_proof.rs

```diff
@@ -0,0 +1,581 @@
+                let sub_path_proof = SubPathProof {
+                    terminal: path_proof.terminal.clone(),
+                    // The depth at which the terminal starts not sharing any siblings
+                    depth: curr_traverse_info.path_bit_index,
+                    inner_siblings: path_proof
+                        .siblings
+                        .iter()
+                        .skip(curr_traverse_info.path_bit_index)
+                        .copied()
+                        .collect(),
+                };
+                sub_paths.push(sub_path_proof);
+
+                // terminal always immediately follows a bisection in a well-formed trie.
+                assert!(ext_sibling_in_sub_tree.is_empty());
+
+                // skip to the next bisection in the stack, if empty we're finished
+                curr_traverse_info = match stack_traverse_info.pop() {
+                    Some(i) => i,
+                    None => break,
+                };
+                continue;
+            }
+
+            // check if at least two key_path in the key_paths slice
+            // has two different bits in position path_index
+            //
+            // they are ordered and all the bits up to path_index
+            // are shared so we can just check the first and the last one differs
+            let path_lower = path_proofs[curr_traverse_info.lower].terminal.key_path();
+            let path_upper = path_proofs[curr_traverse_info.upper].terminal.key_path();
+
+            if path_lower[curr_traverse_info.path_bit_index]
+                != path_upper[curr_traverse_info.path_bit_index]
+            {
+                // if they differ we can skip their siblings but we need to bisect the slice
+                //
+                // binary search between key_paths in the slice to see where to
+                // perform the bisection
+
+                // path_slice_upper and path_slice_lower will never be the same otherwise
+                // bits at path_index would have been the same
+
+                // We have just checked that path_lower and path_upper differ at path_bit_index,
+                // thus, with the vector of path_proofs ordered, we're sure there is at least one
+                // path with its key_path containing a value of 1 at path_bit_index
+                let mid = curr_traverse_info.lower
+                    + path_proofs[curr_traverse_info.lower..=curr_traverse_info.upper]
+                        .binary_search_by(|path_proof| {
+                            if !path_proof.terminal.key_path()[curr_traverse_info.path_bit_index] {
+                                std::cmp::Ordering::Less
+                            } else {
+                                std::cmp::Ordering::Greater
+                            }
+                        })
+                        .expect_err("There must be at least one bit set to 1");
```

Comment:
for a follow-up: codebase style is to prefer `unwrap` with a comment.

Reply:
oh, then I fell behind. Didn't we discuss adding a brief description in `expect` plus a comment explaining why this is impossible? 

In any case, if you made any new decisions, I will change it in a follow-up 👍🏻